Improve hover sounds playback (to reduce volume saturation and delay)

peppy opened the pull request:
This aims to fix two issues:

- Hover sound playback until now was arbitrarily delayed by 50ms, giving them less responsiveness than they should have.
- The debounce interval was per-instance, which basically had zero effect.

The previous 50ms was likely never hit due to being for since instances. Upon making this change, I've readjusted it to 20ms as this seems to feel better to me (50ms feels a bit too... structured? like you can predict when the next sample is going to play).

Considered using a local static dictionary rather than `SessionStatics` but on checking, it would seem such static usage is nowhere in this project and I can guarantee there will be heated discussion if I went in this direction. For now I've just made it shared across all hover sounds types. Still feels fine.

peppy committed:
Reduce default hover sound debounce interval

peppy committed:
Share hover sound debounce across all instances via SessionStatics

bdach commented on osu.Game/Graphics/UserInterface/HoverSounds.cs:
```diff
@@ -24,35 +25,39 @@ public class HoverSounds : CompositeDrawable
         protected override bool OnHover(HoverEvent e)
         {
-            playDelegate?.Cancel();
+            bool requiresDebounce = HoverDebounceTime <= 0;
```
this is backwards, isn't it? also truth be told I don't think this needs to exist, really - always testing against `HoverDebounceTime` even if it's zero/negative should suffice

peppy replied:
Good catch, was added as an afterthought and was broken.

bdach commented on osu.Game/Graphics/UserInterface/HoverSounds.cs:
```diff
@@ -24,35 +25,39 @@ public class HoverSounds : CompositeDrawable
         {
-            playDelegate?.Cancel();
+            bool requiresDebounce = HoverDebounceTime <= 0;
+            bool enoughTimePassedSinceLastPlayback = lastPlaybackTime.Value == 0 || Time.Current - lastPlaybackTime.Value > HoverDebounceTime;
```
checking 0 to determine if a playback has ever happened is sorta sketchy - `Bindable<double?>` could work better? don't really need `Precision` & friends with this one I don't think

peppy replied:
such an edge case i didn't bother. i originally had an inline comment saying i didn't bother. i guess i can try double? but not sure how well bindable handles that..

peppy replied:
`double?`'seems to work as expected, so let's go with that.

peppy committed:
Use nullable doubule to better represent initial playback case

peppy committed:
Remove unnecessary (and broken) requiresDebounce check

bdach committed:
Start with null last hover playback time

bdach committed:
Remove mention of default value in xmldoc

Just bound to get outdated with every change anyway. Look at the actual
default value declaration to see what the default is.

bdach commented:
As I said on discord the `SessionStatics` usage may be contentious but the core idea behind this feature cannot be implemented without having some sort of shared state, so the problem won't go away any time you cut it. And seeing we have that mechanism in place already I see no reason to not use it until it is deemed unfit.
